pkg/controller/ipam/antrea_ipam_controller_test.go

```diff
@@ -147,45 +133,44 @@ func TestReleaseStatefulSet(t *testing.T) {
 	var err error
 	// Wait until pool propagates to the informer
 	pollErr := wait.PollImmediate(100*time.Millisecond, 1*time.Second, func() (bool, error) {
-		allocator, err = poolallocator.NewIPPoolAllocator(testPool, crdClient, poolLister)
+		allocator, err = poolallocator.NewIPPoolAllocator(pool.Name, crdClient, poolLister)
 		if err != nil {
 			return false, nil
 		}
 		return true, nil
 	})
 	require.NoError(t, pollErr)
+	defer allocator.ReleaseStatefulSet(statefulSet.Namespace, statefulSet.Name)
+
+	if int(replicas) < allocator.Total() {
+		// Verify create event was handled by the controller and preallocation was succesfull
+		verifyPoolAllocatedSize(t, pool.Name, poolLister, int(replicas))
+	} else {
+		// Not enough IPs in the pool - preallocation should fail
+		verifyPoolAllocatedSize(t, pool.Name, poolLister, 0)
+	}
 
-	verifyPoolAllocatedSize(t, poolLister, 0)
-
-	// Allocate StatefulSet with dedicated IP Pool
-	err = allocator.AllocateStatefulSet(testNoPool, testWithPool, 5)
-	require.NoError(t, err, "Failed to reserve IPs for StatefulSet")
-	verifyPoolAllocatedSize(t, poolLister, 5)
-
-	// Allocate StatefulSet with namespace-based IP Pool annotation
-	err = allocator.AllocateStatefulSet(testWithPool, testNoPool, 3)
-	require.NoError(t, err, "Failed to reserve IPs for StatefulSet")
-	verifyPoolAllocatedSize(t, poolLister, 8)
+	// Delete StatefulSet
+	k8sClient.AppsV1().StatefulSets(namespace.Name).Delete(context.TODO(), statefulSet.Name, metav1.DeleteOptions{})
 
-	// Delete StatefulSet with namespace-based IP Pool annotation
-	controller.enqueueStatefulSetDeleteEvent(&appsv1.StatefulSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      testNoPool,
-			Namespace: testWithPool,
-		}})
+	// Verify Delete event was processed
+	verifyPoolAllocatedSize(t, pool.Name, poolLister, 0)
+}
 
-	// Verify delete event was handled by the controller
-	verifyPoolAllocatedSize(t, poolLister, 5)
+// This test verifies preallocation of IPs for dedicated IP pool annotation.
+func TestStatefulSetLifecycle_DedicatedPool(t *testing.T) {
+	testStatefulSetLifecycle(t, true, 5)
+}
 
-	// Delete StatefulSet with dedicated IP Pool
-	controller.enqueueStatefulSetDeleteEvent(&appsv1.StatefulSet{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      testWithPool,
-			Namespace: testNoPool,
-		}})
+// This test verifies preallocation of IPs for IP pool annotation based on StatefulSet Namespace.
+func TestStatefulSetLifecycle_NamespacePool(t *testing.T) {
+	testStatefulSetLifecycle(t, false, 7)
+}
 
-	// Verify delete event was handled by the controller
-	verifyPoolAllocatedSize(t, poolLister, 0)
+// This test verifies use case when continuous IP range can not be preallocated.
+// However we don't expect error since preallocation is best-effort feature.
+func TestStatefulSetLifecycle_NoPreallocation(t *testing.T) {
+	testStatefulSetLifecycle(t, false, 20)
```

Comment:
The above tests can be merged to one test in table-driven fashion:
```
func TestStatefulSetLifecycle(t *testing.T) {
	tests := []struct {
		name    string
		dedicatedPool bool
		replicas int32
		expectAllocatedSize int32
	}{
		{
			name:    "Dedicated pool",
			dedicatedPool: true,
			replicas: 5,
			expectAllocatedSize: 5,
		},
		...
		{
			name:    "No preallocation",
			dedicatedPool: false,
			replicas: 20,
			expectAllocatedSize: 0,
		},
	}
	for _, tt := range tests {
		stopCh := make(chan struct{})
		defer close(stopCh)

		namespace, pool, statefulSet := initTestObjects(!tt.dedicatedPool, tt.dedicatedPool, tt.replicas)
		...
		verifyPoolAllocatedSize(t, pool.Name, poolLister, tt.expectAllocatedSize)
		...
		// Verify Delete event was processed
		verifyPoolAllocatedSize(t, pool.Name, poolLister, 0)
	}
}
```

Reply:
Thanks, done